src/overview.md

```diff
@@ -0,0 +1,327 @@
+  `rustc` data structures.
+- We then [_parse_ the stream of tokens][parser] to build an Abstract Syntax
+  Tree (AST).
+- We then take the AST and [convert it to High-Level Intermediate
+  Representation (HIR)][hir]. This is a compiler-friendly representation of the
+  AST.  This involves a lot of desugaring of things like loops and `async fn`.
+- We use the HIR to do [type inference]. This is the process of automatic
+  detection of the type of an expression. **TODO: how `ty` module fits in
+  here**
+- **TODO: Maybe some other things are done here? I think initial type checking
+  happens here? And trait solving?**
+- The HIR is then [lowered to Mid-Level Intermediate Representation (MIR)][mir].
+- The MIR is used for [borrow checking].
+- **TODO: const eval fits in somewhere here I think**
+- We (want to) do [many optimizations on the MIR][mir-opt] because it is still
+  generic and that improves the code we generate later, improving compilation
+  speed too. (**TODO: size optimizations too?**)
+    - MIR is a higher level (and generic) representation, so it is easier to do
+      some optimizations at MIR level than at LLVM-IR level. For example LLVM
+      doesn't seem to be able to optimize the pattern the [`simplify_try`] mir
+      opt looks for.
+- Rust code is _monomorphized_, which means making copies of all the generic
+  code with the type parameters replaced by concrete types. In order to do
+  this, we need to collect a list of what concrete types to generate code for.
+  This is called _monomorphization collection_.
+- We then begin what is vaguely called _code generation_ or _codegen_.
+    - The [code generation stage (codegen)][codegen] is when higher level
+      representations of source are turned into an executable binary. `rustc`
+      uses LLVM for code generation.  The first step is the MIR is then
+      converted to LLVM Intermediate Representation (LLVM IR). This is where
+      the MIR is actually monomorphized, according to the list we created in
+      the previous step.
+    - The LLVM IR is passed to LLVM, which does a lot more optimizations on it.
+      It then emits machine code. It is basically assembly code with additional
+      low-level types and annotations added. (e.g. an ELF object or wasm).
+      **TODO: reference for this section?**
+    - The different libraries/binaries are linked together to produce the final
+      binary. **TODO: reference for this section?**
+
+[`librustc_lexer`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lexer/index.html
+[`librustc_driver`]: https://rust-lang.github.io/rustc-guide/rustc-driver.html
+[lex]: https://rust-lang.github.io/rustc-guide/the-parser.html
+[`StringReader`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_parse/lexer/struct.StringReader.html
+[`librustc_parse`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_parse/index.html
+[parser]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lexer/index.html
```

Comment:
Not sure if we wanted to link to https://doc.rust-lang.org/nightly/nightly-rustc/rustc_parse/index.html here?